File: Enigma/pkg/enigma.go
package pkg

import (
	"math/rand"
	"time"
)

const (
	alphabet = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ"
)

type Reflect struct {
	reflect map[byte]byte
}

type Rotor struct {
	rotorA []byte
	rotorB []byte
	index  int
}

type Enigma struct {
	reflect Reflect
	rotorA  Rotor
	rotorB  Rotor
	rotorC  Rotor
}

func initEnigma(indexA, indeB int) Enigma {
	return Enigma{
		reflect: initReflect(),
		rotorA:  initRotor(indexA),
		rotorB:  initRotor(indexB),
		rotorC:  initRotor(0),
	}
}

func shuffledBytes() []byte {
	rand.Seed(time.Now().UnixNano())

	shuffledBytes := []byte(alphabet)

	for i := len(alphabet) - 1; i > 0; i-- {
		j := rand.Intn(i + 1)
		shuffledBytes[i], shuffledBytes[j] = shuffledBytes[j], shuffledBytes[i]
	}

	return shuffledBytes
}

func initRotor(index int) Rotor {
	return Rotor{
		rotorA: shuffledBytes(),
		rotorB: shuffledBytes(),
		index: index,
	}
}

func initReflect(index int) Reflect {
	a := shuffledBytes()
	b := shuffledBytes()

	reflect := map[byte]byte{}

	for i := 0; i < len(a); i++ {
		reflect[a[i]] = b[i] 
	}

	return Reflect{
		reflect: reflect,
	}
}

File: Enigma/src/enigma.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>
#include "enigma.h"

void shuffle(char *alphabet, int n) {
    for (int i = n - 1; i > 0; i--) {
        int j = rand() % (i + 1);
        char temp = alphabet[i];
        alphabet[i] = alphabet[j];
        alphabet[j] = temp;
    }
}

void init_enigma_rotors(char matrix[ROWS_ROTORS][COLUMNS]) {
    char alphabet[] = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ";
    for (int i = 0; i < ROWS_ROTORS; i++) {
        shuffle(alphabet, COLUMNS);
        for (int j = 0; j < COLUMNS; j++) {
            matrix[i][j] = alphabet[j];
        }
    }
}

void init_enigma_reflect(char matrix[ROWS_REFLECT][COLUMNS]) {
    char alphabet[] = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ";
    shuffle(alphabet, COLUMNS);
    for (int j = 0; j < COLUMNS / 2; j++) {
        matrix[0][j] = alphabet[j];
        matrix[1][j] = alphabet[COLUMNS / 2 + j];
        matrix[0][COLUMNS / 2 + j] = alphabet[COLUMNS / 2 + j];
        matrix[1][COLUMNS / 2 + j] = alphabet[j];
    }
}

void right_rotate(char arr[COLUMNS], int k) {
    k = k % COLUMNS; 
    int temp[k];

    for (int i = 0; i < k; i++) {
        temp[i] = arr[COLUMNS - k + i];
    }

    for (int i = COLUMNS - 1; i >= k; i--) {
        arr[i] = arr[i - k];
    }

    for (int i = 0; i < k; i++) {
        arr[i] = temp[i];
    }
}

int init_enigma(enigma_t *enigma, int indexA, int indexB, int indexC) {
    init_enigma_rotors(enigma->rotors);     //инициализация роторов
    init_enigma_reflect(enigma->reflect);    //инициализация рефлектора

    enigma->indexA = indexA;
    enigma->indexB = indexB; 
    enigma->indexC = indexC;

    right_rotate(enigma->rotors[0], indexA);
    right_rotate(enigma->rotors[2], indexB);
    right_rotate(enigma->rotors[5], indexC);

    return 0;
}
